### PR TITLE
[PB-6085] feature/Added unit tests for photo upload service layer

### PR DESCRIPTION
**Pull request by CandelR**
Covers PhotoBackupFolders caching behaviour, PhotoDeduplicator new/edited asset detection, PhotoDeviceId getOrCreate logic, PhotoUploadQueue success/error/replace flows, and PhotoUploadService URI parsing utilities.                                        
                                                                                                                                                                                                               
  ## Summary                                                                                                                         
  - Unit tests for all services introduced in [Photo backup upload services PR](https://github.com/internxt/drive-mobile/pull/436)
  - Tests covering folder caching, deduplication, device ID persistence, upload queue concurrency, and URI parsing edge cases

### CODE REVIEW

File: src/services/photos/PhotoBackupFolders.spec.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      .mockResolvedValueOnce('month-uuid')
+      .mockResolvedValueOnce('day-15-uuid')
+      .mockResolvedValueOnce('day-20-uuid');
+
+    await photoBackupFolders.getOrCreateFolderForDate(DEVICE_ID, new Date('2024-06-15'));
+    await photoBackupFolders.getOrCreateFolderForDate(DEVICE_ID, new Date('2024-06-20'));
+
+    const folderNames = mockCreateFolder.mock.calls.map((call) => call[1]);
+    const monthCalls = folderNames.filter((name) => name === '06');
+    expect(monthCalls).toHaveLength(1);
```
Comment (TamaraFinogina) [inline src/services/photos/PhotoBackupFolders.spec.ts:59]: shouldn't it also be checked expect(mockCreateFolder).toHaveBeenCalledTimes(9); (5 for May 15 and 4 for May 20)?

File: src/services/photos/PhotoBackupFolders.spec.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  mockCreateFolder
+    .mockResolvedValueOnce('photos-root-uuid')
+    .mockResolvedValueOnce('device-uuid')
+    .mockResolvedValueOnce('year-uuid')
+    .mockResolvedValueOnce('month-uuid')
+    .mockResolvedValueOnce('day-uuid');
+});
+
+describe('PhotoBackupFolders.getOrCreateFolderForDate', () => {
+  test('when called for a date, then creates the full folder hierarchy and returns the day folder uuid', async () => {
```
Comment (TamaraFinogina) [inline src/services/photos/PhotoBackupFolders.spec.ts:31]: Should there be a test for a date in the future? something like: does not create anything for the future date?

File: src/services/photos/PhotoDeviceId.spec.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    mockSetItem.mockResolvedValue(undefined);
+
+    const id = await PhotoDeviceId.getOrCreate();
+
+    expect(id).toBe('generated-uuid');
+    expect(mockSetItem).toHaveBeenCalledTimes(1);
+    expect(mockSetItem).toHaveBeenCalledWith(expect.any(String), 'generated-uuid');
+  });
+
+  test('when no device ID is stored, then successive calls generate only one ID per call', async () => {
```
Comment (TamaraFinogina) [inline src/services/photos/PhotoDeviceId.spec.ts:42]: And if a stored device ID is different? Should we have a test for that too? Say the user has 2 different phones and 1 account

File: src/services/photos/PhotoUploadQueue.spec.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
+import * as MediaLibrary from 'expo-media-library';
+import { AssetUploadJob, PhotoUploadQueue } from './PhotoUploadQueue';
+import { PhotoUploadService } from './PhotoUploadService';
+
+jest.mock('./PhotoUploadService', () => ({
```
Comment (TamaraFinogina) [inline src/services/photos/PhotoUploadQueue.spec.ts:5]: btw, maybe with Vitest directly it will be easier

File: src/services/photos/PhotoUploadService.utils.spec.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  });
+});
+
+describe('stripFileScheme', () => {
+  test('when a file URI is provided, then it returns the decoded path without the scheme', () => {
+    expect(stripFileScheme('file:///data/user/0/com.app/photo.jpg')).toBe('/data/user/0/com.app/photo.jpg');
+  });
+
+  test('when the path contains percent-encoded characters, then it decodes them', () => {
+    expect(stripFileScheme('file:///data/My%20Camera/photo.jpg')).toBe('/data/My Camera/photo.jpg');
```
Comment (TamaraFinogina) [inline src/services/photos/PhotoUploadService.utils.spec.ts:34]: will it work for extension too?

 expect(stripFileScheme('file:///downloads/photo.jpg%2Eexe')).toBe('/downloads/photo.jpg.exe');

File: src/services/photos/PhotoUploadService.utils.spec.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+});
+
+describe('splitFileNameAndExtension', () => {
+  test('when a filename with an extension is provided, then it splits correctly', () => {
+    expect(splitFileNameAndExtension('photo.jpg')).toEqual({ plainName: 'photo', fileExtension: 'jpg' });
+  });
+
+  test('when a filename has multiple dots, then it uses the last dot as the separator', () => {
+    expect(splitFileNameAndExtension('my.vacation.photo.jpg')).toEqual({
+      plainName: 'my.vacation.photo',
```
Comment (TamaraFinogina) [inline src/services/photos/PhotoUploadService.utils.spec.ts:71]: will it do this too? Cause maybe we'll need to blacklist some extensions

expect(splitFileNameAndExtension('my.vacation.photo.apk')).toEqual({
    plainName: 'my.vacation.photo',
    fileExtension: 'apk',
  });

File: src/services/photos/PhotoBackupFolders.spec.ts (hunk 2 above)
Comment (CandelR) [reply src/services/photos/PhotoBackupFolders.spec.ts:31]: The folder is created using the asset's metadata, so if it indicates a future date, it will be created based on that. It may be that the device's time is incorrect, and we shouldn't block the copying of any assets

File: src/services/photos/PhotoDeviceId.spec.ts (hunk 3 above)
Comment (CandelR) [reply src/services/photos/PhotoDeviceId.spec.ts:42]: then it will have 2 ids, the id is for device

File: src/services/photos/PhotoUploadQueue.spec.ts (hunk 4 above)
Comment (CandelR) [reply src/services/photos/PhotoUploadQueue.spec.ts:5]: I’d say that Vitests is specifically designed for Vite, so it’s not the right choice for a React Native project

File: src/services/photos/PhotoUploadService.utils.spec.ts (hunk 6 above)
Comment (CandelR) [reply src/services/photos/PhotoUploadService.utils.spec.ts:71]: This feature only scans media assets. Basically, what you can see in a gallery app

File: src/services/photos/PhotoBackupFolders.spec.ts (hunk 1 above)
Comment (CandelR) [reply src/services/photos/PhotoBackupFolders.spec.ts:59]: We can add it, but there will be `expect(mockCreateFolder).toHaveBeenCalledTimes(6)`, not 9. There’s a cache to prevent unnecessary calls :)

File: src/services/photos/PhotoUploadService.utils.spec.ts (hunk 5 above)
Comment (CandelR) [reply src/services/photos/PhotoUploadService.utils.spec.ts:34]: Yes, it works, and that test could be added. However, this is only used for media assets, with all that that entails in terms of file extensions

**Review by TamaraFinogina — APPROVED** [review]
Sonar complains about duplicated code

**Comment by CandelR** [comment]
> Sonar complains about duplicated code

Yeah, it’s because of the tests. I’ve tried to stop it from checking the duplicate test code, but Sonar just ignores it :(
